### ISSUE TITLE
Forcing an empty array

### ISSUE THREAD

**Issue by shagabutdinov** [thoughtbot/fishery#36]
Hey, thanks for building the fishery, it's awesome! 

I have a question about forcing the empty array of elements:

```
import { Factory } from "fishery";

type Elements = {
  elements: string[];
};

const elements = Factory.define<Elements>(() => ({
  elements: ["STRING"],
}));

console.log(elements.build({ elements: [] }));
// { elements: ["STRING"] }
```

The result of the execution will be non-empty array while I expect the empty array for property `elements`. Is the array does not set to an empty array by design? If it's not expected, mind pointing me where this should be fixed, so I'll be able to open the PR.

Again, thanks for the awesome library, it's absolute pleasure to use it :heart:

**Comment by stevehanson** [comment]
Hi, @shagabutdinov. Thanks for reporting this issue, and glad you're enjoying the library! 

Another [similar issue](https://github.com/thoughtbot/fishery/issues/33) was opened a couple days ago, and the opener ended up closing it since I wasn't able to reproduce. This seems like a real issue though if two people are now reporting it.

Would you mind sharing which version of Fishery you're using, and I'll take a look.

**Comment by stevehanson** [comment]
I just confirmed the behavior you're seeing with a failing test, and I agree that the behavior you expect seems like it should be how this works.

Fishery uses `lodash.merge` to recursively [merge params and associations](https://github.com/thoughtbot/fishery/blob/master/lib/builder.ts#L29) on top of the object returned by the factory. I am seeing the behavior you described in `lodash.merge`:

```javascript
_.merge({ elements: ['STRING'] }, { elements: [] })
// { elements: ['STRING'] }
```

I would expect it to return `{ elements: [] }`. 

It looks like [`lodash.mergeWith`](https://lodash.com/docs/4.17.15#mergeWith) allows passing a customizer to customize the merge behavior. I will look into that to see if there might be a way to get the desired behavior with that.

In the meantime, you should be able to update your factory so that if `elements` is passed to `build()`, it will always take precedence over the factory defaults:

```javascript
const elements = Factory.define<Elements>(({ params }) => ({
  elements: (params.elements as string[]) || ["STRING"],
}));
```

**Comment by shagabutdinov** [comment]
Thanks for the reply! 

I've reproduced the behavior on versions `0.3.0` and `1.0.0`. The suggested fix works, thanks! 

It seems, the behavior that is currently implemented, supports a slightly different case:

```
import { Factory } from "fishery";

type Elements = {
  elements: [string, string];
};

const elements = Factory.define<Elements>(() => ({
  elements: ['STRING1', 'STRING2'],
}));

// works as expected:

console.log(elements.build({ elements: ['STRING3'] }));
// { elements: [ 'STRING3', 'STRING2' ] } 

console.log(elements.build({ elements: ['STRING3', 'STRING4'] }));
// { elements: [ 'STRING3', 'STRING4' ] } 
```

I'm not sure how to correctly implement both `[string, string]` and `string[]` behaviors using typescript.

**Comment by stevehanson** [comment]
As I think more about this, I don't see any reason why the deep merge functionality would need to deep merge tuples or arrays at all. The deep merging was introduced so that we could call `build()` with partial nested objects, eg:

```typescript
const userFactory = Factory.define<User>(() => ({
  name: 'Rosa',
  address: { city: 'Austin', country: 'USA' },
}));

const user = userFactory.build({  address: { city: 'El Paso' } });
user.address.city; // El Paso (from build)
user.address.country; // USA (from factory)
```

But with tuples and arrays, there isn't any merge behavior that I would expect as a user. In fact, I would probably find any merging to be confusing. 

### Tuple example

```typescript
const userFactory = Factory.define<User>(() => ({
  name: 'Rosa',
  twoFavoriteHobbies: ['reading', 'writing'],
}));

// what would a user expect this to do?
userFactory.build({ twoFavoriteHobbies: ['programming'] })
```

In the above example, I think the `build` call should expect `twoFavoriteHobbies` to be the full tuple and generate a compiler error in this case. Anything else would be surprising to me.

### Array example

```typescript
const userFactory = Factory.define<User>(() => ({
  name: 'Rosa',
  posts: [{ id: 1 }, { id: 2 }],
}));

// what would a user expect here?
userFactory.build({ posts: [{ id: 3 }] })
```

In the above, I would expect the passed `posts` to fully override those from the factory (this is the current behavior, we just have that bug with the empty array).

With this in mind, I will work on a fix that short circuits the nested merge when it encounters an array/tuple. I will also see if I can apply the same definition to our `DeepPartial` type so the compiler correctly types arrays so we we get the appropriate compiler errors on `build` and so `params` is typed correctly in factories so we don't have to typecast arrays like we did above in this thread.

Here's also a [relevant lodash issue thread](https://github.com/lodash/lodash/issues/1313), just for my own reference.

**Comment by stevehanson** [comment]
@shagabutdinov I think I put together a solution in #38. I also pulled your tests in. Let me know what you think!

**Comment by stevehanson** [comment]
@shagabutdinov I just wanted to update you that #38 has merged, and I cut a new release, 1.0.1. Thanks again for reporting this and for your help 🌞.

**Comment by shagabutdinov** [comment]
Thanks for the support! :heart: Great to see the good software becomes even better :slightly_smiling_face:
